### PR TITLE
SI-4788 Respect RetentionPolicy of Java classes

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by soc** [comment]
@adriaanm Not finished, but I'd like to get some feedback first.
- [ ] The duplication is really troubling ... I tried to deduplicate some annotation methods, but it caused more trouble than it was worth.
- [ ] Are the additions to `definitions` fine, or should I create those things at a more localized?
- [ ] I'm currently not handling `RetentionPolicy.SOURCE` correctly (ends up as `RuntimeInvisibleAttribute`). That's easy but I haven't found a nice way to implement the logic without computing some tings twice. (I basically need some kind of TriState, but that doesn't play nice with the for-comprehension guard ...).
- [ ] While we are at it, it would be nice to finally get Scala annotations fixed, too.

**Comment by soc** [comment]
Closing. I'm looking at this again when Eugene's PR touching annotations is merged.

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 1]
```diff
@@ -1055,6 +1055,14 @@ trait Definitions extends api.StandardDefinitions {
     lazy val ClassfileAnnotationClass   = requiredClass[scala.annotation.ClassfileAnnotation]
     lazy val StaticAnnotationClass      = requiredClass[scala.annotation.StaticAnnotation]
 
+    lazy val AnnotationRetentionAttr               = requiredClass[java.lang.annotation.Retention]
+    lazy val AnnotationRetentionPolicyAttr         = requiredClass[java.lang.annotation.RetentionPolicy]
+    lazy val AnnotationRetentionPolicySourceValue  = AnnotationRetentionPolicyAttr.children.find(child => child.name == TermName("SOURCE")).get
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]: `getMemberValue(AnnottionRetentionPolicyAttr, TermName("SOURCE"))`?

**Comment by xeno-by** [comment]
My PR only changes the public interface to annotations. Internal workings are left untouched for future work. Therefore I think my changes aren't going to affect your PR, are they?

File: src/reflect/scala/reflect/internal/Definitions.scala (hunk 1 above)
Comment (soc) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]: No, the member doesn't show up this way.

```
val rpv = typeOf[java.lang.annotation.RetentionPolicy].typeSymbol

getMemberValue(rpv, TermName("RUNTIME")) // scala.reflect.internal.FatalError: class RetentionPolicy does not have a member RUNTIME

getMember(rpv, TermName("RUNTIME")) // scala.reflect.internal.FatalError: class RetentionPolicy does not have a member RUNTIME

termMember(rpv, "RUNTIME") // $r.intp.global.Symbol = <none>

getMemberIfDefined(rpv, TermName("RUNTIME")) // $r.intp.global.Symbol = <none>
```

Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]:
```
scala> typeOf[java.lang.annotation.RetentionPolicy].typeSymbol.companionModule.info.member(newTermName("RUNTIME"))
res15: $r.intp.global.Symbol = value RUNTIME
```

Comment (soc) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]: Weird ... it works in the REPL, but `AnnotationRetentionPolicyRuntimeValue` fails with `None.get` when running the tests. WTH?

Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]: The set of sealed subclasses you are accessing in `children` only shows ones that have already been initialized. It depends on the order of typechecking sources and/or forcing the info of symbols from the class files. But we know where to find them: in the members of the companion module.

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 2]
```diff
@@ -978,6 +1026,39 @@ abstract class GenASM extends SubComponent with BytecodeWriters with GenJVMASM {
...
+      !annot.matches(definitions.DeprecatedAttr)
+
+    private def isRuntimeVisible(annot: AnnotationInfo) = {
+      val runtimeValue = definitions.AnnotationRetentionPolicyRuntimeValue
+      annot.atp.typeSymbol.annotations
+        .find(annot => annot.matches(definitions.AnnotationRetentionAttr))
+        .exists(retAnnot => retAnnot.assocs.contains((TermName("value") -> LiteralAnnotArg(Constant(runtimeValue)))))
+    }
+
+    private def retentionPolicyOf(annot: AnnotationInfo): Symbol = {
```
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:1048]: If I add `getCurrentCUnit().warning(annot.symbol.pos, definitions.AnnotationRetentionPolicyAttr.children.toString)` here the result is `Set(value SOURCE, value CLASS, value RUNTIME)`, but in the next line they can't be found anymore?

File: src/reflect/scala/reflect/internal/Definitions.scala (hunk 1 above)
Comment (soc) [inline src/reflect/scala/reflect/internal/Definitions.scala:1060]: Wow, this did it! Thanks for the hint!

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala [hunk 3]
```diff
@@ -702,9 +702,24 @@ abstract class BCodeHelpers extends BCodeTypes with BytecodeWriters {
...
       annot.symbol.initialize.isJavaDefined &&
       annot.matches(definitions.ClassfileAnnotationClass) &&
+      retentionPolicyOf(annot) != definitions.AnnotationRetentionPolicySourceValue &&
       annot.args.isEmpty &&
       !annot.matches(definitions.DeprecatedAttr)
 
+    private def isRuntimeVisible(annot: AnnotationInfo) = {
+      val runtimeValue = definitions.AnnotationRetentionPolicyRuntimeValue
+      annot.atp.typeSymbol.annotations
+        .find(annot => annot.matches(definitions.AnnotationRetentionAttr))
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala:712]: I think there should be a dedicated function for that, something like `annotations.get(AnnotationRetentionAttr)`

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 4]
```diff
@@ -978,6 +1026,39 @@ abstract class GenASM extends SubComponent with BytecodeWriters with GenJVMASM {
...
+      val runtimeValue = definitions.AnnotationRetentionPolicyRuntimeValue
+      val maybeAnnot =
+        annot.atp.typeSymbol.annotations
+          .find(annot => annot.matches(definitions.AnnotationRetentionAttr))
+      val value = maybeAnnot
+        .map(retAnnot => retAnnot.assocs.find(assoc => assoc._1 == TermName("value")).get)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:1054]: nme.value

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 5]
```diff
@@ -978,6 +1026,39 @@ abstract class GenASM extends SubComponent with BytecodeWriters with GenJVMASM {
...
+     *   .initialize: if 'annot' is read from pickle, atp might be un-initialized
+     *
+     * must-single-thread
+     */
+    private def shouldEmitAnnotation(annot: AnnotationInfo) =
+      annot.symbol.initialize.isJavaDefined &&
+      annot.matches(definitions.ClassfileAnnotationClass) &&
+      retentionPolicyOf(annot) != definitions.AnnotationRetentionPolicySourceValue &&
+      annot.args.isEmpty &&
+      !annot.matches(definitions.DeprecatedAttr)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:1039]: So the idea is to never emit Scala annotations, right?
\

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:1039]: Also, isn't isJavaDefived guaranteeing that it's a subclass of ClassfileAnnotation?

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:1039]: Oh interesting, so DeprecatedAttr is java-defined?

**Pull request closed without merging**